Fix alpha CRD openapi e2e

roycaihw opened the pull request:
/kind failing-test

**What this PR does / why we need it**:
make schema structural so it can be published

```
- lastTransitionTime: "2019-05-21T22:37:25Z"
   message: '[spec.validation.openAPIV3Schema.properties[spec].properties[dummy].type:
     Required value: must not be empty for specified object fields, spec.validation.openAPIV3Schema.properties[status].properties[bars].type:
     Required value: must not be empty for specified object fields]'
   reason: Violations
   status: "True"
   type: NonStructuralSchema
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref #78087 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

/cc @liggitt @sttts 
/test pull-kubernetes-e2e-gce-alpha-features

roycaihw committed:
fix crd openapi alpha e2e

make schema used in e2e structural so it can be published;
GET before UPDATE to tolerate object change

liggitt commented:
Looks like the e2e test doesn't tolerate conflict errors (which it needs to since updates trigger the controller modifying the condition). Wrapping the update in a RetryOnConflict would resolve that error

liggitt commented:
looks like `works for CRD without validation schema` is also failing unrelatedly:

> error: error validating "STDIN": error validating data: ValidationError(E2e-test-crd-publish-openapi-8455-crd): unknown field "a" in io.k8s.crd-publish-openapi-test-empty.v1.E2e-test-crd-publish-openapi-8455-crd; if you choose to ignore these errors, turn validation off with --validate=false

can you compare what got published in openapi before and after the structural schema openapi PR? it seems likely what we're publishing now could break kubectl

roycaihw commented:
> can you compare what got published in openapi before and after the structural schema openapi PR? it seems likely what we're publishing now could break kubectl

working on it

roycaihw committed:
restore test behavior for CRD without validation cases

the test doesn't set empty validation but expects CRD controller to treat
nil schema specially and publish an empty schema

roycaihw commented:
/test pull-kubernetes-e2e-gce-alpha-features

liggitt commented:
/test pull-kubernetes-e2e-gce-alpha-features

liggitt commented:
/test pull-kubernetes-e2e-gce-alpha-features

roycaihw commented:
/retest

liggitt commented:
/lgtm
/approve
/priority critical-urgent

roycaihw commented:
/area custom-resources
